### PR TITLE
#17396 Support TiDB

### PR DESCRIPTION
**Pull request by Icemap**
Hi guys. I add TiDB support for DBeaver. 
[TiDB](https://docs.pingcap.com/tidb/stable/overview) is an open-source NewSQL database.
Many of our customers want to use TiDB with DBeaver.

### CODE REVIEW

**Comment by Icemap** [comment]
PTAL @LonwoLonwo @serge-rider

**Comment by E1izabeth** [comment]
Thank you for contribution. Please, revert all unintended changes in the localization files.

**Comment by Icemap** [comment]
Hi, @E1izabeth . I can revert this localization files. But those files changed by plugin: [ResourceBundle Editor](http://essiembre.github.io/eclipse-rbe/). Should I revert and change those manually?

**Comment by E1izabeth** [comment]
Yes

**Comment by Icemap** [comment]
Got it.

**Comment by Icemap** [comment]
@E1izabeth Completed.

File: plugins/org.jkiss.dbeaver.ext.mysql.ui/src/org/jkiss/dbeaver/ext/mysql/ui/views/MySQLConnectionPage.java [hunk 1]
```diff
@@ -72,14 +74,18 @@ public void dispose()
         super.dispose();
         UIUtils.dispose(LOGO_MYSQL);
         UIUtils.dispose(LOGO_MARIADB);
+        UIUtils.dispose(LOGO_TIDB);
     }
 
     @Override
     public Image getImage() {
+    	String driverID = getSite().getDriver().getId();
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.mysql.ui/src/org/jkiss/dbeaver/ext/mysql/ui/views/MySQLConnectionPage.java]: Please correct indentation.

File: plugins/org.jkiss.dbeaver.ext.mysql.ui/src/org/jkiss/dbeaver/ext/mysql/ui/views/MySQLConnectionPage.java [hunk 2]
```diff
@@ -72,14 +74,18 @@ public void dispose()
...
         // We set image only once at activation
         // There is a bug in Eclipse which leads to SWTException after wizard image change
-        if (getSite().getDriver().getId().equalsIgnoreCase(MySQLConstants.DRIVER_ID_MARIA_DB)) {
+        if (driverID.equalsIgnoreCase(MySQLConstants.DRIVER_ID_MARIA_DB)) {
             return LOGO_MARIADB;
+        } else if (driverID.equalsIgnoreCase(MySQLConstants.DRIVER_ID_TIDB)) {
+        	return LOGO_TIDB;
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.mysql.ui/src/org/jkiss/dbeaver/ext/mysql/ui/views/MySQLConnectionPage.java]: Please correct indentation.

File: plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/MySQLUtils.java [hunk 3]
```diff
@@ -147,7 +147,20 @@ public static boolean isMariaDB(DBPDriver driver) {
                 driver.getDriverClassName());
     }
 
-    public static boolean isAlterUSerSupported(MySQLDataSource dataSource) {
-        return dataSource.isMariaDB() ? dataSource.isServerVersionAtLeast(10, 2) : dataSource.isServerVersionAtLeast(5, 7);
+    public static boolean isTiDB(DBPDriver driver) {
+    	return MySQLConstants.DRIVER_ID_TIDB.equals(driver.getId());
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/MySQLUtils.java]: and here

File: plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/MySQLUtils.java [hunk 4]
```diff
@@ -147,7 +147,20 @@ public static boolean isMariaDB(DBPDriver driver) {
...
+    	if (dataSource.isTiDB()) {
+    		// TiDB always support
+    		return true;
+    	} else if (dataSource.isMariaDB()) {
+    		// MariaDB 10.2+
+    		return dataSource.isServerVersionAtLeast(10, 2);
+    	} else {
+    		// MySQL 5.7+
+    		return dataSource.isServerVersionAtLeast(5, 7);
+    	}
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/MySQLUtils.java]: and here too

File: plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java [hunk 5]
```diff
@@ -792,7 +795,50 @@ public boolean isMariaDB() {
...
+    		
+    		// target major version is less than server 
+    		// OR
+    		// major version equals AND target minor version is equal or less than server
+    		return (major < serverMajor || (major == serverMajor && minor <= serverMinor));
+    	} catch (NumberFormatException e){
+    		// It means not a TiDB server actually
+    		log.error("TiDB server version not a number, error: " + e.getMessage());
+    		return false;
+    	}
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java]: All this block should be 4 spaces indented.

File: plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java [hunk 6]
```diff
@@ -822,10 +868,13 @@ public ErrorPosition[] getErrorPosition(@NotNull DBRProgressMonitor monitor, @No
     }
 
     public boolean supportsCheckConstraints() {
-        if (this.isMariaDB()) {
+    	if (this.isTiDB()) {
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java]: the same for this block of code

File: plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java [hunk 7]
```diff
@@ -842,18 +891,38 @@ public boolean supportsInformationSchema() {
...
+    	} else if (this.isMariaDB()) {
             return this.isServerVersionAtLeast(10, 3);
         }
+
         return false;
     }
 
+	// TiDB: not support triggers
+    public boolean supportsTriggers() {
+    	return !this.isTiDB();
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java]: and here

File: plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java [hunk 8]
```diff
@@ -842,18 +891,38 @@ public boolean supportsInformationSchema() {
...
+    }
+
+	// TiDB: not support procedures
+    public boolean supportsProcedures() {
+    	return !this.isTiDB();
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java]: and here

File: plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java [hunk 9]
```diff
@@ -842,18 +891,38 @@ public boolean supportsInformationSchema() {
...
+    }
+    
+	// TiDB: not support event
+    public boolean supportsEvents() {
+    	return !this.isTiDB();
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java]: and here

File: plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/plan/MySQLPlanAnalyser.java [hunk 10]
```diff
@@ -66,8 +66,12 @@ public MySQLPlanAbstract explain(JDBCSession session, String query) throws DBCEx
         }
     }
 
+    // MariaDB 10.1+ & MySQL 5.6+: support
+    // rest MariaDB/MySQL & TiDB: not support
     private boolean supportsExplainJSON() {
-        if (dataSource.isMariaDB()) {
+    	if (dataSource.isTiDB()) {
+    		return false;
+    	} else if (dataSource.isMariaDB()) {
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/plan/MySQLPlanAnalyser.java]: and here

**Review by E1izabeth — CHANGES_REQUESTED** [review]
There are some issues according the code style. Please correct indentation.

File: plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java (hunk 5 above)
Comment (Icemap) [reply plugins/org.jkiss.dbeaver.ext.mysql/src/org/jkiss/dbeaver/ext/mysql/model/MySQLDataSource.java]: OK, I'll replace all tabs to 4 spaces.

**Comment by Icemap** [comment]
@E1izabeth Completed. 

BTW, I found the CI pipeline emerged with an [error](https://github.com/dbeaver/dbeaver/runs/7915882681?check_suite_focus=true). It seems I should create a ticket at JIRA and rebase or force push to change those commit messages.
May I know the JIRA URL to create a ticket?

**Comment by E1izabeth** [comment]
We use github issues numbers for commit messages. So you can just create a feature request on the github.

**Comment by Icemap** [comment]
OK, ref to this feature request: https://github.com/dbeaver/dbeaver/issues/17396

**Review by ShadelessFox — APPROVED** [review]

**Review by E1izabeth — APPROVED** [review]

**Comment by E1izabeth** [comment]
We passed your pull request to our QA. Waiting for testing.

**Comment by Icemap** [comment]
THX. If this PR has any bugs, I'll fix them ASAP.

**Comment by serge-rider** [comment]
Generally I don't like that it is in the main MySQL/Maria bundle.
It will make us to take care of TiDB  every time we change or add something to MySQL bundle.

The good colution would be to create a new extension org.jkiss.dbeaver.ext.tidb and extend MySQL classes there.
Some changes may be needed in the base mysql bundle but only a minor ones.

@Icemap WDYT?

**Comment by dveeden** [comment]
I had a quick look at this PR:

Merging `devel` into this branch resulted in one minor conflict that was easily fixed.

I had to make one small change to get this to work:
```patch

 diff --git a/plugins/org.jkiss.dbeaver.ext.mysql/plugin.xml b/plugins/org.jkiss.dbeaver.ext.mysql/plugin.xml
index 87de100972..4d67600400 100644
--- a/plugins/org.jkiss.dbeaver.ext.mysql/plugin.xml
+++ b/plugins/org.jkiss.dbeaver.ext.mysql/plugin.xml
@@ -280,7 +280,7 @@
                     promoted="1"
                     categories="sql">
 
-                    <file type="jar" path="maven:/mysql:mysql-connector-java:RELEASE" bundle="!drivers.mysql"/>
+                    <file type="jar" path="maven:/mysql:mysql-connector-java:RELEASE[8.0.29]" bundle="!drivers.mysql"/>
                     <file type="license" path="drivers/mysql/LICENSE.txt" bundle="drivers.mysql"/>
                     <file type="jar" path="drivers/mysql/mysql8" bundle="drivers.mysql"/>
 
```

The functionality looks good to me. The only know functionality known not to work is explain:

image

The code was updated not to use `MySQLPlanJSON` if the server is TiDB, so it now uses `MySQLPlanClassic`, but that tries to execute "EXPLAIN EXTENDED...", which results in a SQLError in TiDB due to the `EXTENDED` keyword that TiDB doesn't support. Note that removing the `EXTENDED` keyword is not sufficient to make this work as the column names etc. of the TiDB explain format are different. See https://docs.pingcap.com/tidb/dev/sql-statement-explain for some examples.

With MySQL, TiDB, MariaDB and in the future maybe something like https://github.com/vitessio/vitess I can see that splitting might make sense. However I would expect that most differences would be between MySQL and MariaDB as they are both following their own path. TiDB is expected to stay close to the syntax, drivers, etc. that MySQL uses. Not sure which option is the best in the long term.

**Comment by Icemap** [comment]
Alright. @serge-rider @E1izabeth 

I'll add two modules `org.jkiss.dbeaver.ext.tidb` and `org.jkiss.dbeaver.ext.tidb.ui` to support TiDB. 
And prepared based on `org.jkiss.dbeaver.ext.oceanbase` and `org.jkiss.dbeaver.ext.oceanbase.ui` to build the TiDB support.

BTW, Mac OS has some weird errors to [develop DBeaver](https://github.com/dbeaver/dbeaver/discussions/17981) now. Currently, I only can using Windows to develop it.

**Comment by Icemap** [comment]
Let's using this PR: https://github.com/dbeaver/dbeaver/pull/18299

**Pull request closed without merging**